File: python/agnsf/variability.py
"""Single-curve variability statistics (user-facing Python layer).

All numbers are computed by the C++ core (`agnsf::variability`); this
module only wraps the result:

    result = agnsf.variability.measure(t, v, e)
    print(result.fvar, result.fvar_uncertainty)
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Sequence

from ._agnsf import (
    Uncertainty,
    variability_measure as _variability_measure,
)

__all__ = ["measure", "VariabilityResult"]


@dataclass
class VariabilityResult:
    """Variability statistics of one light curve.

    Definitions follow Vaughan et al. (2003); ``F_var`` (fractional
    variability amplitude), ``nxs`` (normalized excess variance) and
    ``xs`` (excess variance) carry analytic uncertainties from the
    same reference (``None`` when not defined).
    """

    n: int
    valid: bool
    mean: float
    weighted_mean: float
    weighted_mean_error: float
    stddev: float
    stddev_error: float
    peak_to_peak: float
    peak_to_peak_noise_corrected: float
    sigma_m: float
    fvar: float
    fvar_uncertainty: Uncertainty | None
    nxs: float
    nxs_uncertainty: Uncertainty | None
    xs: float
    xs_uncertainty: Uncertainty | None
    chi2: float
    chi2_dof: float
    chi2_q: float
    von_neumann: float

    @property
    def fvar_error(self) -> float:
        """Symmetric uncertainty of ``fvar`` (NaN when not estimated)."""
        if self.fvar_uncertainty is None:
            return float("nan")
        return 0.5 * (self.fvar_uncertainty.upper - self.fvar_uncertainty.lower)

    @property
    def nxs_error(self) -> float:
        if self.nxs_uncertainty is None:
            return float("nan")
        return 0.5 * (self.nxs_uncertainty.upper - self.nxs_uncertainty.lower)

    @property
    def xs_error(self) -> float:
        if self.xs_uncertainty is None:
            return float("nan")
        return 0.5 * (self.xs_uncertainty.upper - self.xs_uncertainty.lower)


def measure(
    time: Sequence[float],
    value: Sequence[float],
    error: Sequence[float],
    *,
    err_sys: float = 0.0,
    weighted: bool = False,
) -> VariabilityResult:
    """Compute single-curve variability statistics.

    Parameters
    ----------
    time, value, error:
        Observed time, flux, and per-point uncertainty arrays.
    err_sys:
        Systematic error floor added in quadrature when subtracting
        the noise contribution.
    weighted:
        Use the inverse-variance weighted mean inside the variance /
        intrinsic-amplitude estimates.
    """
    raw = _variability_measure(
        time,
        value,
        error,
        float(err_sys),
        bool(weighted),
    )

    return VariabilityResult(
        n=raw.n,
        valid=raw.valid,
        mean=raw.mean,
        weighted_mean=raw.weighted_mean,
        weighted_mean_error=raw.weighted_mean_error,
        stddev=raw.stddev,
        stddev_error=raw.stddev_error,
        peak_to_peak=raw.peak_to_peak,
        peak_to_peak_noise_corrected=raw.peak_to_peak_noise_corrected,
        sigma_m=raw.sigma_m,
        fvar=raw.fvar,
        fvar_uncertainty=raw.fvar_uncertainty,
        nxs=raw.nxs,
        nxs_uncertainty=raw.nxs_uncertainty,
        xs=raw.xs,
        xs_uncertainty=raw.xs_uncertainty,
        chi2=raw.chi2,
        chi2_dof=raw.chi2_dof,
        chi2_q=raw.chi2_q,
        von_neumann=raw.von_neumann,
    )

File: python/bindings/variability.cpp
#include "common.hpp"
#include <variability/variability.hpp>

using namespace agnsf::python;


// ------------------------------------------------------------------
// Single-curve variability statistics
// ------------------------------------------------------------------

void bind_variability(py::module_& m)
{
    py::class_<agnsf::variability::Statistics>(
        m,
        "VariabilityResult"
    )
    .def_readonly(
        "n",
        &agnsf::variability::Statistics::n
    )
    .def_readonly(
        "valid",
        &agnsf::variability::Statistics::valid
    )
    .def_readonly(
        "mean",
        &agnsf::variability::Statistics::mean
    )
    .def_readonly(
        "weighted_mean",
        &agnsf::variability::Statistics::weighted_mean
    )
    .def_readonly(
        "weighted_mean_error",
        &agnsf::variability::Statistics::weighted_mean_error
    )
    .def_readonly(
        "stddev",
        &agnsf::variability::Statistics::stddev
    )
    .def_readonly(
        "stddev_error",
        &agnsf::variability::Statistics::stddev_error
    )
    .def_readonly(
        "peak_to_peak",
        &agnsf::variability::Statistics::peak_to_peak
    )
    .def_readonly(
        "peak_to_peak_noise_corrected",
        &agnsf::variability::Statistics::peak_to_peak_noise_corrected
    )
    .def_readonly(
        "sigma_m",
        &agnsf::variability::Statistics::sigma_m
    )
    .def_readonly(
        "fvar",
        &agnsf::variability::Statistics::fvar
    )
    .def_readonly(
        "fvar_uncertainty",
        &agnsf::variability::Statistics::fvar_uncertainty
    )
    .def_readonly(
        "nxs",
        &agnsf::variability::Statistics::nxs
    )
    .def_readonly(
        "nxs_uncertainty",
        &agnsf::variability::Statistics::nxs_uncertainty
    )
    .def_readonly(
        "xs",
        &agnsf::variability::Statistics::xs
    )
    .def_readonly(
        "xs_uncertainty",
        &agnsf::variability::Statistics::xs_uncertainty
    )
    .def_readonly(
        "chi2",
        &agnsf::variability::Statistics::chi2
    )
    .def_readonly(
        "chi2_dof",
        &agnsf::variability::Statistics::chi2_dof
    )
    .def_readonly(
        "chi2_q",
        &agnsf::variability::Statistics::chi2_q
    )
    .def_readonly(
        "von_neumann",
        &agnsf::variability::Statistics::von_neumann
    );

    m.def(
        "variability_measure",
        [](
            const Float64Array& time,
            const Float64Array& value,
            const Float64Array& error,
            double err_sys,
            bool weighted
        )
        {
            const agnsf::LightCurveView curve =
                make_light_curve_view(time, value, error);

            agnsf::variability::Options options;
            options.err_sys = err_sys;
            options.weighted = weighted;

            return agnsf::variability::measure(curve, options);
        },
        py::arg("time"),
        py::arg("value"),
        py::arg("error"),
        py::arg("err_sys") = 0.0,
        py::arg("weighted") = false,
        R"pbdoc(
Compute single-curve variability statistics.

Returns a VariabilityResult with the weighted/plain mean, sample
standard deviation, peak-to-peak amplitude, intrinsic sigma_m,
fractional variability amplitude F_var, normalized and
unnormalized excess variance (with Vaughan et al. 2003 analytic
uncertainties), the chi^2 variability significance Q, and the von
Neumann ratio.

err_sys: systematic error floor added in quadrature when the noise
            contribution is subtracted.
weighted: use the inverse-variance weighted mean inside the
            variance estimates.
        )pbdoc"
    );
}
